Source/Core/Core/IOS/IOSC.h

```diff
@@ -0,0 +1,127 @@
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+// Implementation of an IOSC-like API, but much simpler since we only support actual keys.
+
+#pragma once
+
+#include <array>
+#include <map>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+#include "Common/Crypto/AES.h"
+
+class PointerWrap;
+
+namespace IOS
+{
+namespace HLE
+{
+enum ReturnCode : s32;
+
+class IOSC final
+{
+public:
+  IOSC();
+  ~IOSC();
+
+  using Handle = u32;
+  // We use the same default key handle IDs as the actual IOSC because there are ioctlvs
+  // that accept arbitrary key handles from the PPC, so the IDs must match.
+  // More information on default handles: https://wiibrew.org/wiki/IOS/Syscalls
+  enum DefaultHandle : u32
+  {
+    // ECC-233 private signing key (per-console)
+    HANDLE_CONSOLE_KEY = 0,
+    // Console ID
+    HANDLE_CONSOLE_ID = 1,
+    // NAND FS AES-128 key
+    HANDLE_FS_KEY = 2,
+    // NAND FS HMAC
+    HANDLE_FS_MAC = 3,
+    // Common key
+    HANDLE_COMMON_KEY = 4,
+    // PRNG seed
+    HANDLE_PRNG_KEY = 5,
+    // SD AES-128 key
+    HANDLE_SD_KEY = 6,
+    // boot2 version (writable)
+    HANDLE_BOOT2_VERSION = 7,
+    // Unknown
+    HANDLE_UNKNOWN_8 = 8,
+    // Unknown
+    HANDLE_UNKNOWN_9 = 9,
+    // Filesystem version (writable)
+    HANDLE_FS_VERSION = 10,
+    // New common key (aka Korean common key)
+    HANDLE_NEW_COMMON_KEY = 11,
+  };
+
+  enum ObjectType : u8
+  {
+    TYPE_SECRET_KEY = 0,
+    TYPE_PUBLIC_KEY = 1,
+    TYPE_DATA = 3,
+  };
+
+  enum ObjectSubType : u8
+  {
+    SUBTYPE_AES128 = 0,
+    SUBTYPE_MAC = 1,
+    SUBTYPE_ECC233 = 4,
+    SUBTYPE_DATA = 5,
+    SUBTYPE_VERSION = 6
+  };
+
+  // Create an object for use with the other functions that operate on objects.
+  ReturnCode CreateObject(Handle* handle, ObjectType type, ObjectSubType subtype, u32 pid);
+  // Delete an object. Built-in objects cannot be deleted.
+  ReturnCode DeleteObject(Handle handle, u32 pid);
+  // Import a secret, encrypted key into dest_handle, which will be decrypted using decrypt_handle.
+  ReturnCode ImportSecretKey(Handle dest_handle, Handle decrypt_handle, u8* iv,
+                             const u8* encrypted_key, u32 pid);
+  // Import a public key.
+  ReturnCode ImportPublicKey(Handle dest_handle, const u8* public_key, u32 pid);
+  // Compute an AES key from an ECDH shared secret.
+  ReturnCode ComputeSharedKey(Handle dest_handle, Handle private_handle, Handle public_handle,
+                              u32 pid);
+
+  // AES encrypt/decrypt.
+  ReturnCode Encrypt(Handle key_handle, u8* iv, const u8* input, size_t size, u8* output,
+                     u32 pid) const;
+  ReturnCode Decrypt(Handle key_handle, u8* iv, const u8* input, size_t size, u8* output,
+                     u32 pid) const;
+
+  // Ownership
+  ReturnCode GetOwnership(Handle handle, u32* owner) const;
+  ReturnCode SetOwnership(Handle handle, u32 owner, u32 pid);
+
+  void DoState(PointerWrap& p);
```

Comment:
Good catch, thanks! Kernel now calls DoState and I have also bumped the version.